tests/unit/test_charm.py

```diff
@@ -1,18 +1,674 @@
+        self, context
+    ):
+        state = State(
+            storage=[Storage(name="config"), Storage(name="database")],
+            containers=[Container(name="gocert", can_connect=False)],
+            networks={"juju-info": Network([], [], [])},
+            leader=True,
+        )
+
+        with patch("requests.get", return_value=Mock(side_effect=requests.ConnectionError)):
+            out = context.run(Event("config-changed"), state)
+
+        assert out.unit_status == ops.WaitingStatus("container not yet connectable")
+
+    def test_given_only_config_storage_container_can_connect_network_not_available_gocert_not_running_when_configure_then_status_is_waiting(
+        self, context
+    ):
+        state = State(
+            storage=[Storage(name="config")],
+            containers=[Container(name="gocert", can_connect=True)],
+            networks={"juju-info": Network([], [], [])},
+            leader=True,
+        )
+
+        with patch("requests.get", return_value=Mock(side_effect=requests.ConnectionError)):
+            out = context.run(Event("config-changed"), state)
+        assert out.unit_status == ops.WaitingStatus("storages not yet available")
+
+    def test_given_only_database_storage_container_can_connect_network_not_available_gocert_not_running_when_configure_then_status_is_waiting(
+        self, context
+    ):
+        state = State(
+            storage=[Storage(name="database")],
+            containers=[Container(name="gocert", can_connect=True)],
+            networks={"juju-info": Network([], [], [])},
+            leader=True,
+        )
+
+        with patch("requests.get", return_value=Mock(side_effect=requests.ConnectionError)):
+            out = context.run(Event("config-changed"), state)
+        assert out.unit_status == ops.WaitingStatus("storages not yet available")
+
+    def test_given_storages_available_container_can_connect_network_not_available_gocert_not_running_when_configure_then_config_file_generated(
+        self, context
+    ):
+        state = State(
+            storage=[Storage(name="config"), Storage(name="database")],
+            containers=[Container(name="gocert", can_connect=True)],
+            networks={"juju-info": Network([], [], [])},
+            leader=True,
+        )
+
+        with patch("requests.get", return_value=Mock(side_effect=requests.ConnectionError)):
+            out = context.run(Event("config-changed"), state)
+        assert len(out.secrets) == 0
+        root = out.containers[0].get_filesystem(context)
+        assert (root / "var/lib/gocert/config/config.yaml").open("r")
+        assert not (root / "var/lib/gocert/config/certificate.pem").exists()
+        assert not ((root / "var/lib/gocert/config/private_key.pem").exists())
+        assert out.unit_status == ops.WaitingStatus("certificates not yet created")
```

Comment:
I would avoid having a test that asserts both the content of files and unit status as those are handled by two different handlers ( though I know the collect unit status runs as an exit hook to other events).

Reply:
changed